[Marcus Tan] iP

emptygx opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind from having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,

1. download it from [here](https://github.com/emptygx/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)
---
If you are Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public static void main(String[] args) {
    new Duke("./data/duke.txt").run();
}
```
---
Thank you for using `Duke` 😄 

```
 ____        _        
|  _ \ _   _| | _ ___ 
| | | | | | | |/ / _ \
| |_| | |_| |   <  __/
|____/ \__,_|_|\_\___|
```

Commit:
Add Gradle support

emptygx committed:
Level-1

emptygx committed:
Level 2: Add, List

emptygx committed:
Added task class

emptygx committed:
Level 3. Mark as Done

emptygx committed:
Created subclasses of task

emptygx committed:
Level 4: ToDos, Events, Deadlines

emptygx committed:
Level 5: Handle Errors

emptygx committed:
Level 6: Delete

emptygx committed:
improve DukeException.java

emptygx committed:
Duke class: clean up implementation of exceptions

emptygx committed:
Revert "improve DukeException.java"

This reverts commit 4b08b8eeb94c966202501443ef2ec746df4c2d78.

emptygx committed:
Revert "Revert "improve DukeException.java""

This reverts commit bfe990725e63d6630398112976ee5285b7ed78fe.

emptygx committed:
Duke class: add enums for different cases

emptygx committed:
reformat case branch indentation

emptygx committed:
Add FileSaver.java and file saving capabilities for Duke

emptygx committed:
Ignore files in data/

emptygx committed:
Add support for dates and times

emptygx committed:
Merge branch 'branch-Level-8'

emptygx committed:
Remove printing of file and directory existence status

emptygx committed:
Merge branch 'branch-test'

emptygx committed:
Add TaskList.java to encapsulate ArrayList of tasks

emptygx committed:
Rename FileSaver.java to Storage.java and update capabilities

emptygx committed:
Add abstract class Command.java

emptygx committed:
Add command subclasses AddCommand.java, DeleteCommand.java, DoneCommand.java, ExitCommand.java, ListCommand.java, and UndoneCommand.java

emptygx committed:
Add Ui.java to handle UI of program

emptygx committed:
Add Parser.java to handle parsing of user input

emptygx committed:
Update program code to incorporate more OOP and better functionality

emptygx committed:
Put classes in a package named "duke"

emptygx committed:
Add package statement for all files in duke

emptygx committed:
Remove comment block

emptygx committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

emptygx committed:
Configure checkstyle for gradle

emptygx committed:
Add Junit tests in DukeTest.java, ParserTest.java and TaskTest.java and edit files to support testing

emptygx committed:
Create a JAR file and edit Duke welcome print text

emptygx committed:
Add JavaDoc headers to classes

emptygx committed:
Tweak code to comply with java coding standard

emptygx committed:
Add FindCommand.java and update classes to include search capabilities

emptygx committed:
Merge branch 'branch-A-CodingStandard'

emptygx committed:
Merge branch 'branch-Level-9'

emptygx committed:
Add JavaDoc comments to several public methods

emptygx committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/DukeException.java

weiquany commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+    private TaskList tasks;
+    private Ui ui;
+    /**
+     * Initializes the application.
```
Should there be an empty line between the description and parameter section for the Javadoc comment? Similarly for Javadoc comments in other parts of the application.

weiquany commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,62 @@
+        }
+        String description = split.length == 2 ? split[1] : "";
+        switch (command) {
+        case BYE:
```
Good job following the switch case indentation convention of CS2103!

weiquany commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,49 @@
+    private final PrintStream originalOut = System.out;
+    private final PrintStream originalErr = System.err;
+
+    public void setUpStreams() {
```
This is a very useful way to set up tests! I have not tried it myself but I think there may be a way to automate it, rather than doing it manually in each of the test cases. You can try adding `@BeforeEach` and `@AfterEach`. If you tried and it works or does not work, please feel free to let me know!

weiquany reviewed (commented):
Great job overall on adhering to the CS2103 coding standards! I had a difficult time looking for any possible mistakes, and may have made mistakes in my review, do feel free to correct me!

Kok-je commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,103 @@
+
+    private String filePath;
+    /**
+     * Creates a new storage that stores the task list in the given file.
```
perhaps could consider adding an empty line in between the description and paras section

Kok-je commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+            String output = length == 1 ? " task in the list." : " tasks in the list.";
+            System.out.println("Got it. I've deleted this task:\n" + t
+                    + "\nNow you have " + length + output);
+        } catch (IndexOutOfBoundsException e) {
```
Can consider encapsulating index out of bounds as a Duke exception here, and just trigger the Duke exception instead of sout the string message? But nonetheless, great effort in being so meticulous with all the possible error cases!

Kok-je commented on src/main/java/duke/DoneCommand.java:
```diff
@@ -0,0 +1,23 @@
+ * Marks a task as done.
+ */
+public class DoneCommand extends Command {
+    private int index;
```
Can considering leaving lines in between here for readability

Kok-je commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public class Event extends Task {
+
+    private LocalDate at;
```
Hi Marcus, it was not immediately obvious what at meant. I would like to suggest using a more descriptive name, such as startTime.

Kok-je commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,84 @@
+     * @param task the task that was added
+     * @param size the size of the TaskList after adding the task
+     */
+    public void showAddTask(Task task, int size) {
```
Just a small point for consideration, separating the operands to + in System.out.println() and aligning them neatly can give a better visualisation of what is meant to be printed in one quick glance.

Kok-je reviewed (commented):
Great job Marcus! Overall, the code was very easy to read because the indentation and whitespace rules are observed very carefully. Just made a few minor comments on the code that you can look through!

emptygx committed:
Implement basic GUI using JavaFX

emptygx committed:
Fix some coding standard violations

emptygx committed:
Fix width of DialogContainer in GUI

emptygx committed:
Change program entry point to Launcher.java

emptygx committed:
Implement Java varargs

emptygx committed:
Main class: make the GUI window a fixed size

GUI window is resizable.

Resizing the GUI causes issues to layout of elements. A fixed size window is more
user-friendly as it prevents unnecessary confusion.

Let's disable resizing for the GUI stage and for cosmetic purposes set the title to "Duke".

emptygx committed:
Use assertions feature in code

Assertions not used in code.

Important assumptions made about the code at various points can be verified and
drastic action can be taken if the assumptions hold false.

Let's insert assertions at various points in the code.

emptygx committed:
Fix exit command to close GUI window

emptygx committed:
Merge branch 'branch-Level-10'

emptygx committed:
Merge branch 'master' into branch-A-Assertions

emptygx committed:
Merge branch 'master' into branch-A-CodeQuality

emptygx committed:
Improve code quality

emptygx committed:
Use the streams feature of Java in TaskList class

emptygx committed:
Merge pull request #2 from emptygx/branch-A-Assertions

branch-A-Assertions

emptygx committed:
Merge branch 'master' of https://github.com/emptygx/ip

emptygx committed:
Merge branch 'master' into branch-A-CodeQuality

emptygx committed:
Merge branch 'master' into branch-A-Streams

emptygx committed:
Merge pull request #3 from emptygx/branch-A-CodeQuality

branch-A-CodeQuality

emptygx committed:
Merge branch 'master' of https://github.com/emptygx/ip

emptygx committed:
Merge branch 'master' into branch-A-Streams

emptygx committed:
Merge pull request #4 from emptygx/branch-A-Streams

branch-A-Streams

emptygx committed:
Merge branch 'master' of https://github.com/emptygx/ip

emptygx committed:
Add gradle.yml to set up CI

emptygx committed:
Add new assign priority to task feature

emptygx committed:
Merge pull request #5 from emptygx/branch-C-Priority

Add new assign priority to task feature

emptygx committed:
Merge branch 'master' of https://github.com/emptygx/ip

emptygx committed:
Improve GUI, add personality, change Duke's name to TASKY

emptygx committed:
Improve GUI dialog box design

emptygx committed:
Add more javadoc comments

emptygx committed:
Update UI responses to add personality

emptygx committed:
Add representative screenshot of product

emptygx committed:
Update readme to include all features and product details
